lsp/server/src/validateDocument.ts

```diff
@@ -25,25 +26,26 @@ export async function validateDocument(
     document: TextDocument,
     extensionName: string
 ): Promise<Diagnostic[]> {
-    const { uri } = document;
-
-    const setting = await getDocumentSettings(uri);
+    const setting = await getDocumentSettings(document.uri);
     let results: Diagnostic[] = [];
 
     if (document.languageId === 'javascript') {
         // handles JS rules
         const jsDiagnostics = await validateJs(document);
-        
+
         // handle graphql rules
         const graphqlDiagnostics = await validateGraphql(document);
 
         results = results.concat(jsDiagnostics, graphqlDiagnostics);
-        results.splice(setting.maxNumberOfProblems);
     }
 
     if (document.languageId === 'html') {
+        const mobileOfflineDiagnostics = await validateMobileOffline(document);
```

Comment:
It would be better if we have a `validateHTML(document: Document')` and a separate `mobile-offline-tag` rule under `diagnostic/html` folder. It will open the door for future diagnostic suppressing. Also it would be easier to maintain in case we will add more rule to html files

Reply:
How much do you want this? I don't mind making the change. But I have to note that the way diagnostics is setup for JS and graphQL is that rules are implementing interface `DiagnosticProducer<T>` which itself declares `validateDocument(...)` . The parameter takes `astNode:T` which is designed to parse AST. Whereas, for html element, the single parameter is `TextDocument`. We can force to make the validators scaffold alike but will soon find out that they're different.

Another argument, and less convincing, I would make is that since this is the one and only html validator still I think future proofing like this isn't strongly required.

That is why I ask, how much do you want this to be done?

Reply:
param `astNode:T` within `validateDocument` is confusing. It should be able to handle any date structure since it is generic. `TextDocument` should be fine.  It is better to put `mobile-offline` as ` rule since we take take the advantage of `DiagnosticProducer` to support suppressing

Reply:
That's what I'm saying. Validating for mobile offline DOESN'T USE DiagnosticProducer. I'm just asking to make sure we're not mashing up things for the sake of everything need to follow the same approach.

Reply:
What I am suggesting is to use `DiagnosticProducer` for html files to keep the pattern the same. We can change the signature of `DiagnosticProducer` interface to export interface DiagnosticProducer<T> {
    ```
/**
     * Validate the parsed text document as astNode and return a list of diagnostics.
     * @param textDocument the language server text document
     * @param data  data parsed from document body
     * @returns An array of diagnostics found within document
     */
    validateDocument(
        textDocument: TextDocument,
        data: T
    ): Promise<Diagnostic[]>;
}

```

Reply:
But mobile offline validator doesn't need `data:T`. That's someone else's concern. You're trying to mashup method signature when there is no need for the sake of trying to look alike.

Reply:
The need is for future suppressing support and each `DiagnosticProducer` is going to have an ID for fine tuning validator suppression. For example, in `setting.json`, we can define `“mobile.lsp.diagnostic”: {
            “suppress.all”: “true”,
            “suppressed.rule.Ids”: [“adapters-local-change-not-aware”, "mobile-offline"],
            “maxProblemNumber”: 500
 }`

If you don't like to create another `DiagnosticProducer`, we can come up with other approach to support it

Reply:
it will be great to has validateHtml.ts and move the offline individual into diagnostic rule folder.  this moment it's end of world.  I can do the re-arrangement when I will merge your change into the suppressing piece I am doing. 

so for this one, take as is?

Reply:
@haifeng-li-at-salesforce where is the rule ID defined? I don't see it. If there's a valid reason as you stated, like I said, I'm not against changing. But with the way your code has been written I don't know how `"suppressed.rule.Ids"` are supposed to work. Is this something you were planning on doing it later?

Reply:
yes, I am working on the suppressing stuff now, hope to have the pr tomorrow morning.  we can huddle and show what the implementation.

Reply:
I would never be able to know what you and/or Haifeng are preparing now unless it is already in the codebase. And to be asked for something that's not there is harder. Let's huddle tomorrow. If I'm understanding you correctly there's a work you're working on now that should come before this(my) PR then.

Reply:
Hi @sfdctaka I am sorry I commented before showing the diagnostic pr.  now the [preview pr](https://github.com/salesforce/salesforcedx-vscode-mobile/pull/135) is up.